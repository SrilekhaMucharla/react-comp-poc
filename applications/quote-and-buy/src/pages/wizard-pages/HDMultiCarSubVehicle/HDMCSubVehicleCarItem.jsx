/* eslint-disable no-else-return */
import React, { useState, useEffect } from 'react';
import { Col, Row } from 'react-bootstrap';
import PropTypes from 'prop-types';
import _ from 'lodash';
import {
    HDLabelRefactor, HDQuoteInfoRefactor
} from 'hastings-components';
import {
    AnalyticsHDToggleButtonGroup as HDToggleButtonGroup
} from '../../../web-analytics';
import formatRegNumber from '../../../common/formatRegNumber';
import * as messages from './HDMultiCarSubVehicle.messages';

const HDMotorLegalCarItem = ({
    mcObj, index, subvehiclebuttonHandle, mcsubmissionVM, ancillaryJourneyDataModel, isProvisional, invalidateImportantStuffPage
}) => {
    const [type, settype] = useState('');
    const [registrationNumber, setregistrationNumber] = useState('');
    const [amount, setamount] = useState(0);
    const [foundAncillary, setfoundAncillary] = useState(null);
    const [car, setcar] = useState('');
    const makePath = 'lobData.privateCar.coverables.vehicles[0].make';
    const modelPath = 'lobData.privateCar.coverables.vehicles[0].model';
    const coveragePath = 'coverages.privateCar.ancillaryCoverages[0].coverages';
    const availableValues = [{
        value: 'yes',
        name: messages.yes,
    }, {
        value: 'no',
        name: messages.no,
    }];

    const addonHandler = (event, idx) => {
        invalidateImportantStuffPage();
        subvehiclebuttonHandle(event.target.value, idx, amount, registrationNumber, type);
    };

    useEffect(() => {
        if (mcObj) {
            let regNum = mcsubmissionVM[index].lobData.privateCar.coverables.vehicles[0].license;
            // eslint-disable-next-line prefer-template
            regNum = formatRegNumber(regNum);
            setregistrationNumber(regNum);
            let foundAncillary1 = [];
            if (_.get(mcObj, coveragePath)) {
                const brandCover = _.get(mcObj, coveragePath);
                foundAncillary1 = brandCover.filter((item) => {
                    return item.name === messages.pageTitleText;
                });
                const tempAmt = _.get(foundAncillary1[0], 'amount.amount', 0);
                const amt = `Â£${tempAmt.toFixed(2)}`;
                setamount(foundAncillary1.length > 0 ? amt : null);
            }
            const make = _.get(mcsubmissionVM[index], makePath);
            const model = _.get(mcsubmissionVM[index], modelPath);
            const tmpCar = `${make} ${model}`;
            setcar(tmpCar);
            setfoundAncillary(foundAncillary1);
            settype(mcObj.quote.branchCode);
        }
    }, [mcObj]);

    const checkancillarymodalSelection = (type) => {
        let btn = '';
        if (ancillaryJourneyDataModel && ancillaryJourneyDataModel.substituteVehicle.length) {
            ancillaryJourneyDataModel.substituteVehicle.forEach((mc) => {
                if (mc.quoteID === mcObj.quoteID) {
                    if (foundAncillary && foundAncillary.length && foundAncillary[0].selected) {
                        btn = 'yes';
                    } else if (foundAncillary && foundAncillary.length && foundAncillary[0].selected === false && mc.substituteVehicle) {
                        btn = 'no';
                    }
                }
            });
            return btn;
        } else {
            return null;
        }
    };

    return (
        <>
            <Row className="align-items-end">
                <Col>
                    <HDLabelRefactor
                        Tag="span"
                        className="reg-num--anc-car-item margin-bottom-tiny"
                        text={registrationNumber} />
                    <HDLabelRefactor
                        Tag="h3"
                        text={car}
                        className="mb-0" />
                </Col>
                <Col xs="auto">
                    {amount !== null && <HDLabelRefactor Tag="h2" text={amount} className="my-0" />}
                </Col>
            </Row>
            <HDToggleButtonGroup
                webAnalyticsEvent={{ event_action: `${messages.ancillaries} - ${messages.pageTitle}` }}
                id="mc-substitute-vehicle-button-group"
                btnGroupClassName="grid grid--col-2 grid--col-lg-3 margin-top-md"
                btnClassName={isProvisional ? 'bg-light text-secondary border-secondary' : 'theme-white'}
                disabled={isProvisional}
                availableValues={availableValues}
                // eslint-disable-next-line no-nested-ternary
                value={checkancillarymodalSelection(messages.HD)}
                onChange={(e) => { addonHandler(e, index); }} />
            {isProvisional && (
                <HDQuoteInfoRefactor>
                    <span>{messages.infoTipText}</span>
                </HDQuoteInfoRefactor>
            )}
        </>
    );
};

HDMotorLegalCarItem.propTypes = {
    mcObj: PropTypes.shape({
        quote: PropTypes.object,
        coverType: PropTypes.string,
        coverages: PropTypes.object,
        insurancePaymentType: PropTypes.string,
        ncdgrantedProtectionInd: PropTypes.bool,
        ncdgrantedYears: PropTypes.string,
        otherOfferedQuotes: PropTypes.object,
        periodEndDate: PropTypes.object,
        periodStartDate: PropTypes.object,
        producerCode: PropTypes.string,
        quoteID: PropTypes.string,
        sessionUUID: PropTypes.string,
        voluntaryExcess: PropTypes.string

    }).isRequired,
    mcsubmissionVM: PropTypes.shape({
        baseData: PropTypes.object,
        bindData: PropTypes.object,
        lobData: PropTypes.object,
        quoteData: PropTypes.object,
        isParentPolicy: PropTypes.string,
        quoteID: PropTypes.bool
    }).isRequired,
    index: PropTypes.number.isRequired,
    subvehiclebuttonHandle: PropTypes.func.isRequired,
    ancillaryJourneyDataModel: PropTypes.shape({
        substituteVehicle: PropTypes.object
    }).isRequired,
    isProvisional: PropTypes.bool.isRequired
};

export default HDMotorLegalCarItem;
