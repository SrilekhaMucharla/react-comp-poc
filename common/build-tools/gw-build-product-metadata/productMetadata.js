const _ = require('lodash');
const { readJson } = require('fs-extra');
const appPaths = require('gw-build-config-paths');
const { sendJsonRpcRequest } = require('gw-build-jsonrpc');
const fileUtils = require('./util/fileUtils');
const metadataLocalisation = require('./metadataLocalisation');

function toMetadataMap(serverMetadataResult) {
    const metadataByName = _.groupBy(serverMetadataResult, 'name');
    return _.mapValues(metadataByName, (typeMetas) => {
        const allMetadataForName = _.map(typeMetas, 'metadata');
        return _.merge({}, ...allMetadataForName);
    });
}

/**
 * Fetches the metadata for the given endpoint
 *
 * @param {ApplicationEndpoint} endpoint the endpoint which metadata should be fetched
 * @returns {Promise<Object>} the metadata retrieved from the given endpoint
 */
function fetchMetadataForEndpoint(endpoint) {
    return sendJsonRpcRequest(endpoint, 'getMetaData')
        .then((metadata) => ({
            endpoint,
            metadata
        }));
}

/**
 * @typedef { import("gw-build-capabilities-management").ApplicationEndpoint } ApplicationEndpoint
 */

/**
 * @typedef PreprocessMetadata
 * @property {{server: string, service: string}} endpoint the server nickname (e.g. pc, cc, ...)
 * @property {object} metadata the metadata from the server
 */

/**
 * Structures the product metadata
 *
 * @param {PreprocessMetadata} data containing the endpoint from which
 * the metadata came from and the metadata from the server
 * @returns {Object} the metadata retrieved from the given endpoint
 */
function preprocessDtoMetadata({ endpoint, metadata }) {
    return {
        [endpoint.server]: toMetadataMap(metadata.newTypeMeta)
    };
}

/**
 * Structures the question set metadata
 *
 * @param {PreprocessMetadata} data containing the endpoint from which
 * the metadata came from and the metadata from the server
 * @returns {Object} the metadata from the server
 */
function preprocessQuestionSetsMetadata({ metadata }) {
    return metadata.questionSets;
}

/**
 * Combines all the metadatas from all the endpoints into one object
 *
 * @param {Array<Object>} allMetadataResponses containing the metadata from the server
 * @returns {Object} the metadata to save
 */
function mergeMetadata(allMetadataResponses) {
    return allMetadataResponses.reduce((generatedMetadata, serviceMetadata) => {
        return _.merge(generatedMetadata, serviceMetadata);
    }, {});
}

/**
 * Pipeline to format, combine and save the metadata
 *
 * @param {Promise<Object>} metadataPromises response from the server
 * @param {function} preprocessingFunction function to format the metadata from the server
 * @param {string} destinationFile path to save the metadata
 * @returns {Object} the metadata to save
 */
function metadataPipeline(metadataPromises, preprocessingFunction, destinationFile) {
    return metadataPromises
        .then((metadataFromEndpoints) => metadataFromEndpoints.map(preprocessingFunction))
        .then(mergeMetadata)
        .then((allMetadata) => fileUtils.writeFile(allMetadata, destinationFile));
}

/**
 * Process the product metadata
 *
 * @param {Promise<Object>} metadataPromises response from the server
 * @returns {Promise} the metadata to save
 */
function processDTOMetadata(metadataPromises) {
    return metadataPipeline(metadataPromises, preprocessDtoMetadata, appPaths.appProductMetadata);
}

/**
 * Process the question sets metadata
 *
 * @param {Promise<Object>} metadataPromises response from the server
 * @returns {Promise} the metadata to save
 */
function processQuestionSetsMetadata(metadataPromises) {
    return metadataPipeline(metadataPromises, preprocessQuestionSetsMetadata,
        appPaths.appQuestionSetsMetadata);
}

async function buildMetadata() {
    const { capabilities } = await readJson(appPaths.appGeneratedConfig);
    const endpoints = Object.values(capabilities);
    const endpointsWithMetadata = endpoints.filter((endpoint) => !endpoint.skipMetadata);
    const metadataPromises = Promise.all(endpointsWithMetadata.map(fetchMetadataForEndpoint));
    await Promise.all([
        processDTOMetadata(metadataPromises),
        processQuestionSetsMetadata(metadataPromises)
    ]);
    const localeConfig = await metadataLocalisation.fetchLocaleInfo();
    metadataLocalisation.processMetadataDisplayKeys(localeConfig, endpoints);
}

module.exports = {
    buildMetadata
};
