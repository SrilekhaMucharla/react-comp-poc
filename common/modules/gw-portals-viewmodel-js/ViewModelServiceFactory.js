import InsuranceSuiteMetadataServiceFactory from 'gw-portals-metadata-js';
import DateUtil from 'gw-portals-util-js/DateUtil';
import LocalDateUtil from 'gw-portals-util-js/LocalDateUtil';

import ViewModel from './ViewModel';

import Context from './aspects/Context';
import Requiredness from './aspects/Requiredness';
import Validity from './aspects/Validity';
import SubtreeValidity from './aspects/SubtreeValidity';
import InputCtrlType from './aspects/InputCtrlType';
import AvailableValues from './aspects/AvailableValues';
import SerializedForm from './aspects/SerializedForm';
import RebuildVMFromData from './aspects/RebuildVMFromData';
import DebugInvalidChildProperties from './aspects/DebugInvalidChildProperties';
import TypedValue from './aspects/TypedValue';
import _ValidationFunctions from './aspects/_ValidationFunctions';
import _AddressAutocompleteUtil from './aspects/_AddressAutocompleteUtil';

import { ExpressionLanguageServiceFactory } from './expression-language';
import _PhoneUtil from './util/PhoneUtil';
import DateFormatter from './util/DateFormatter';

function getViewModelService(metadata, translator) {
    const suiteMetadata = InsuranceSuiteMetadataServiceFactory.createSuiteMetadata(metadata);
    const ExpressionLanguageService = ExpressionLanguageServiceFactory.get(
        { insuranceSuiteMetadataService: suiteMetadata, translator }
    );
    // view model initiation which is done in the initial module in Portals (extracted)
    const ViewModelService = ViewModel.get(suiteMetadata);

    const typeValidationMessages = {};
    typeValidationMessages.notANumber = translator('platform.modelvalidation.DTOValidationService.Please enter numbers only');
    typeValidationMessages.notAnInteger = translator('platform.modelvalidation.DTOValidationService.Value entered must not contain decimal values');

    ViewModelService.addAspectFactory(Context.create(ExpressionLanguageService));
    ViewModelService.addAspectFactory(Requiredness.create(ExpressionLanguageService));
    ViewModelService.addAspectFactory(
        Validity.create(ExpressionLanguageService, typeValidationMessages)
    );
    ViewModelService.addAspectFactory(SubtreeValidity);
    ViewModelService.addAspectFactory(InputCtrlType);
    ViewModelService.addAspectFactory(AvailableValues.create(ExpressionLanguageService));
    ViewModelService.addAspectFactory(SerializedForm);
    ViewModelService.addAspectFactory(RebuildVMFromData);
    ViewModelService.addAspectFactory(DebugInvalidChildProperties);
    ViewModelService.addAspectFactory(TypedValue);

    ExpressionLanguageService.getCompilationContext().registerType('edgev10.aspects.validation.ValidationFunctions', _ValidationFunctions());
    ExpressionLanguageService.getCompilationContext().registerType('gw.api.contact.AddressAutocompleteUtil', _AddressAutocompleteUtil());
    ExpressionLanguageService.getCompilationContext().registerType('edgev10.capabilities.locale.util.DateFormatter', DateFormatter);
    ExpressionLanguageService.getCompilationContext().registerType('gw.api.util.DateUtil', DateUtil);
    ExpressionLanguageService.getCompilationContext().registerType('gw.api.util.PhoneUtil', _PhoneUtil());
    ExpressionLanguageService.getCompilationContext().registerType('edgev10.time.LocalDateUtil', LocalDateUtil);

    return {
        create: ViewModelService.create,
        changeContext: ViewModelService.changeContext,
        clone: ViewModelService.clone
    };
}

// EXPORT
export default {
    getViewModelService
};
