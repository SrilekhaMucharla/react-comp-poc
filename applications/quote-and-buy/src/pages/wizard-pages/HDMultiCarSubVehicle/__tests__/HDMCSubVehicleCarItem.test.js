import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import {
    HDLabelRefactor
} from 'hastings-components';
import ViewModelServiceFactory from 'gw-portals-viewmodel-js/ViewModelServiceFactory';
import productMetadata from '../../../../generated/metadata/product-metadata.json';
import defaultTranslator from '../../__helpers__/testHelper';
import HDMCSubVehicleCarItem from '../HDMCSubVehicleCarItem';

const middlewares = [];
const mockStore = configureStore(middlewares);
describe('<HDManualAddresssPopup />', () => {
    let wrapper;
    let store;
    const submissionVM = {
        value: {
            quotes: [
                {
                    baseData: { brandCode: 'PCW', },
                    lobData: {
                        privateCar: {
                            offerings: [{
                                branchName: 'HastingsPremium',
                                branchCode: 'HP',
                                coverages: [
                                    {
                                        name: 'Motor Legal Expense',
                                        selected: false
                                    }
                                ]
                            }]
                        }
                    }
                }
            ]
        }
    };

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
            wrapper = undefined;
        }
    });

    beforeEach(() => {
        const viewModelService = ViewModelServiceFactory.getViewModelService(
            productMetadata, defaultTranslator
        );
        const mcsubmissionVM = viewModelService.create(
            submissionVM,
            'pc',
            'edgev10.capabilities.quote.submission.dto.QuoteDataDTO'
        );
        const initialState = {
            wizardState: {
                data: {
                    submissionVM: mcsubmissionVM
                },
                app: {
                    step: 1,
                    prevStep: 0
                },
            }
        };
        store = mockStore(initialState);
    });

    test('render component', () => {
        const emptyStore = mockStore({});
        wrapper = shallow(
            <Provider store={emptyStore}>
                <HDMCSubVehicleCarItem mcsubmissionVM={submissionVM} />
            </Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should contain HDLabel', async () => {
        wrapper = mount(
            <Provider store={store}>
                <HDMCSubVehicleCarItem mcsubmissionVM={submissionVM} />
            </Provider>
        );
        expect(wrapper.find(HDLabelRefactor).exists()).toBe(true);
    });
});
