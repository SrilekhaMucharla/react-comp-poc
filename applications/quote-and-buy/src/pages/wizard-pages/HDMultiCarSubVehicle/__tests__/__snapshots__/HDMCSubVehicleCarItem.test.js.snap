// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HDManualAddresssPopup /> render component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <HDMotorLegalCarItem
    mcsubmissionVM={
      Object {
        "value": Object {
          "quotes": Array [
            Object {
              "baseData": Object {
                "brandCode": "PCW",
              },
              "lobData": Object {
                "privateCar": Object {
                  "offerings": Array [
                    Object {
                      "branchCode": "HP",
                      "branchName": "HastingsPremium",
                      "coverages": Array [
                        Object {
                          "name": "Motor Legal Expense",
                          "selected": false,
                        },
                      ],
                    },
                  ],
                },
              },
            },
          ],
        },
      }
    }
  />
</ContextProvider>
`;
